File: jeongmin/스택_큐/주식가격.py
# 처음 푼 풀이
def solution(prices):
    answer = []
    stock_length = len(prices)
    for now_idx in range(stock_length):
        if now_idx == stock_length-1:
            answer.append(0)
            return answer
        else:
            count = 0
            for next_idx in range(now_idx + 1, stock_length):
                count += 1
                if prices[now_idx] > prices[next_idx]:
                    break
            answer.append(count)
    return answer


# 스택으로 푸는 풀이!!
# def solution(prices):
#     n = len(prices)
#     # 답을 넣을 공간 크기에 맞게 미리 만들어놓기(정적 할당)
#     answer = [0] * n
#     stack = []
#
#     # 인덱스 비교를 위해 전체 순회한다
#     for i in range(n):
#         # 스택이 비어 있지 않고(앞 가격들이 기준 가격보다 더 작은 것들이 있을때)
#         # 현재 가격이 스택의 마지막 가격보다 낮으면(스택마지막부터 비교해서 일자 늘려나가는 것)
#         while stack and prices[i] < prices[stack[-1]]:
#             # 제일 가까운 날의 가격 인덱스
#             idx = stack.pop()
#             # 다음날보다 가격이 높은 거니까 가격이 떨어져있는 상태일때 일자 측정(현재와 과거 인덱스=일자 비교)
#             answer[idx] = i - idx
#         # 당일 가격 인덱스 넣기(비교용으로 나보다 작은 가격 나오기 전까지 담겨있게 된다)
#         stack.append(i)
#
#     # 스택에 남아있는 인덱스 처리 (끝까지 가격이 떨어지지 않음)
#     while stack:
#         # 최종적으로 나보다 작은 가격이 나오지 않은 것들의 인덱스
#         idx = stack.pop()
#         # 전체 날에서 들어간 날 빼기
#         answer[idx] = n - 1 - idx
#
#     return answer

File: 프로그래머스/2/42584. 주식가격/주식가격.java
import java.util.*;
class Solution {
    public int[] solution(int[] prices) {
        int priceSize = prices.length;
        int[] answer = new int[priceSize];
        Stack<Integer> stack = new Stack();
        
        // 스택에 인덱스를 넣는다!
        // 만약 값이 작아지는 시점이 등장한다면? -> 전의 인덱스 값을 빼고 현재 인덱스와의 차이 값을 넣는다. 
        for(int i = 0; i < priceSize; i++){
            while(!stack.isEmpty()&&prices[i]<prices[stack.peek()]){
                int value = stack.pop();
                answer[value] = i - value;
            }
            stack.push(i);
        }
        
        // 마지막까지 스택이 남는다면? 
        // 마지막 인덱스에서 뺀 인덱스까지 가격이 떨어지지 않은 것
        while(!stack.isEmpty()){
            int temp = stack.pop();
            answer[temp] = priceSize - temp - 1; 
        }
        
        
        
        // Queue<Integer> q = new ArrayDeque<>();
        // for(int token : prices){
        //     q.add(token);
        // }
        // //while문 안에 빼서 list에 값들 하나씩 비교 
        // ArrayList<Integer> answerList = new ArrayList<>();
        // int[] answer = new int[prices.length];
        // boolean [] isChecked = new boolean[prices.length];
        // int tmp = q.poll();
        // answerList.add(tmp);
        // while(!q.isEmpty()){
        //     int temp = q.poll();
        //     answerList.add(temp);
        //     for(int i = 0; i<answerList.size()-1; i++){
        //         int value = answerList.get(i);
        //         if(!isChecked[i]){
        //             if(temp>=value){ //주식 가격이 오른 경우 
        //                 answer[i]++;
        //             }else{//주식 가격이 내려간 경우 
        //                 answer[i]++;
        //                 isChecked[i] = true; 
        //             }   
        //         }
        //     }
        // }
        return answer;
    }
}